src/hotspot/share/opto/library_call.cpp

```diff
@@ -1202,16 +1202,32 @@ bool LibraryCallKit::inline_string_indexOf(StrIntrinsicNode::ArgEnc ae) {
   Node* tgt_start = array_element_address(tgt, intcon(0), T_BYTE);
   Node* tgt_count = load_array_length(tgt);
 
-  if (ae == StrIntrinsicNode::UU || ae == StrIntrinsicNode::UL) {
-    // Divide src size by 2 if String is UTF16 encoded
-    src_count = _gvn.transform(new RShiftINode(src_count, intcon(1)));
-  }
-  if (ae == StrIntrinsicNode::UU) {
-    // Divide substring size by 2 if String is UTF16 encoded
-    tgt_count = _gvn.transform(new RShiftINode(tgt_count, intcon(1)));
+  Node* result = nullptr;
+  bool do_intrinsic =
+      (StubRoutines::string_indexof() != nullptr) &&
+      ((ae == StrIntrinsicNode::LL) || (ae == StrIntrinsicNode::UU));
+
+  if (!do_intrinsic) {
+    if (ae == StrIntrinsicNode::UU || ae == StrIntrinsicNode::UL) {
+      // Divide src size by 2 if String is UTF16 encoded
+      src_count = _gvn.transform(new RShiftINode(src_count, intcon(1)));
+    }
+    if (ae == StrIntrinsicNode::UU) {
+      // Divide substring size by 2 if String is UTF16 encoded
+      tgt_count = _gvn.transform(new RShiftINode(tgt_count, intcon(1)));
+    }
   }
 
-  Node* result = make_indexOf_node(src_start, src_count, tgt_start, tgt_count, result_rgn, result_phi, ae);
+  if (do_intrinsic) {
+    Node* call = make_runtime_call(RC_LEAF, OptoRuntime::string_IndexOf_Type(),
+                                   StubRoutines::string_indexof(),
+                                   "stringIndexOf", TypePtr::BOTTOM, src_start,
+                                   src_count, tgt_start, tgt_count);
+    result = _gvn.transform(new ProjNode(call, TypeFunc::Parms));
+  } else {
+    result = make_indexOf_node(src_start, src_count, tgt_start, tgt_count,
```

Comment:
Existing routines emits IR to handle following special cases.

tgt_cnt > src_cnt return -1
tgt_cnt == 0 return 0.

Should we not be preserving those check before calling stub ?

As of now these checks are part of stub and doing them in JIT code will save call overhead.